src/oic/oauth2/message.py

```diff
@@ -160,26 +161,32 @@ def to_urlencoded(self, lev=0):
 
         _spec = self.c_param
         if not self.lax:
-            for attribute, (_, req, _ser, _, _) in _spec.items():
-                if req and attribute not in self._dict:
+            for attribute, cparam in _spec.items():
+                if cparam.required and attribute not in self._dict:
                     raise MissingRequiredAttribute("%s" % attribute,
                                                    "%s" % self)
 
         params = []
 
         for key, val in self._dict.items():
             try:
```

Comment:
The whole nested try/except could be done like this:

    cparam = _spec.get(key, _spec.get(key.split('#')[0], _spec.get('*'))

Not sure if thats any better.

Nearly the same pattern (just differs in the finally part) happens in line 261 and 310, and very similar in line 355, just written differently  so maybe a small helper function would be good?
e.g.

    def extract_cparam(_spec, key):
        for _key in (key, key.split('#')[0], '*'):
            if _key in _spec:
                return _spec[key]